File: backend/user-service.js
const { pool } = require('./db');
const bcrypt = require('bcryptjs');

class UserService {
  // ── GET ALL GC USERS/CLIENTS ─────────────────────────────────────────

  static async getAllClients(limit = 50, offset = 0) {
    try {
      const result = await pool.query(
        `SELECT u.id, u.email, u.name, u.company, u.created_at,
                COUNT(DISTINCT p.id) as project_count,
                cs.active_modules,
                cs.created_at as subscription_date
         FROM users u
         LEFT JOIN projects p ON u.id = p.user_id
         LEFT JOIN client_subscriptions cs ON u.id = cs.client_id
         GROUP BY u.id, cs.id
         ORDER BY u.created_at DESC
         LIMIT $1 OFFSET $2`,
        [limit, offset]
      );

      const totalResult = await pool.query('SELECT COUNT(*) as count FROM users');
      
      return {
        success: true,
        clients: result.rows,
        total: parseInt(totalResult.rows[0].count)
      };
    } catch (err) {
      console.error('Get clients error:', err);
      return { success: false, error: err.message };
    }
  }

  // ── GET CLIENT BY ID ─────────────────────────────────────────────────

  static async getClient(clientId) {
    try {
      const result = await pool.query(
        `SELECT u.id, u.email, u.name, u.company, u.created_at,
                COUNT(DISTINCT p.id) as project_count,
                cs.active_modules
         FROM users u
         LEFT JOIN projects p ON u.id = p.user_id
         LEFT JOIN client_subscriptions cs ON u.id = cs.client_id
         WHERE u.id = $1
         GROUP BY u.id, cs.id`,
        [clientId]
      );

      if (result.rows.length === 0) {
        return { success: false, error: 'Client not found' };
      }

      return { success: true, client: result.rows[0] };
    } catch (err) {
      console.error('Get client error:', err);
      return { success: false, error: err.message };
    }
  }

  // ── GET ALL ADMIN USERS ──────────────────────────────────────────────

  static async getAllAdminUsers(limit = 50, offset = 0) {
    try {
      const result = await pool.query(
        `SELECT a.id, a.email, a.admin_level, a.client_id, a.permissions,
                a.is_active, a.created_at, a.last_login,
                u.name as client_name, u.company as client_company
         FROM admin_users a
         LEFT JOIN users u ON a.client_id = u.id
         ORDER BY a.created_at DESC
         LIMIT $1 OFFSET $2`,
        [limit, offset]
      );

      const totalResult = await pool.query('SELECT COUNT(*) as count FROM admin_users');
      
      return {
        success: true,
        admins: result.rows,
        total: parseInt(totalResult.rows[0].count)
      };
    } catch (err) {
      console.error('Get admin users error:', err);
      return { success: false, error: err.message };
    }
  }

  // ── CREATE ADMIN USER ────────────────────────────────────────────────

  static async createAdminUser(adminData) {
    try {
      const { email, password, admin_level, client_id, permissions } = adminData;

      if (!email || !password || !admin_level) {
        return { success: false, error: 'Email, password, and admin_level required' };
      }

      const validLevels = ['super_admin', 'client_admin'];
      if (!validLevels.includes(admin_level)) {
        return { success: false, error: 'Invalid admin_level' };
      }

      // Hash password
      const passwordHash = await bcrypt.hash(password, 12);

      // If client_admin, require client_id
      if (admin_level === 'client_admin' && !client_id) {
        return { success: false, error: 'client_id required for client_admin' };
      }

      const result = await pool.query(
        `INSERT INTO admin_users (email, password_hash, admin_level, client_id, permissions, is_active)
         VALUES ($1, $2, $3, $4, $5, true)
         RETURNING id, email, admin_level, client_id, permissions, is_active, created_at`,
        [email, passwordHash, admin_level, client_id || null, JSON.stringify(permissions || { pricing: true, features: true, users: true })]
      );

      console.log('✓ Admin user created:', result.rows[0].email);
      return { success: true, admin: result.rows[0] };
    } catch (err) {
      if (err.code === '23505') {
        return { success: false, error: 'Email already exists' };
      }
      console.error('Create admin user error:', err);
      return { success: false, error: err.message };
    }
  }

  // ── UPDATE ADMIN USER ────────────────────────────────────────────────

  static async updateAdminUser(adminId, updates) {
    try {
      const allowedFields = ['is_active', 'permissions'];
      const updateFields = [];
      const updateValues = [];
      let paramCount = 1;

      for (const [key, value] of Object.entries(updates)) {
        if (allowedFields.includes(key)) {
          updateFields.push(`${key} = $${paramCount}`);
          updateValues.push(key === 'permissions' ? JSON.stringify(value) : value);
          paramCount++;
        }
      }

      if (updateFields.length === 0) {
        return { success: false, error: 'No valid fields to update' };
      }

      updateValues.push(adminId);
      const query = `UPDATE admin_users SET ${updateFields.join(', ')}, updated_at = NOW() WHERE id = $${paramCount} RETURNING *`;

      const result = await pool.query(query, updateValues);

      if (result.rows.length === 0) {
        return { success: false, error: 'Admin user not found' };
      }

      console.log('✓ Admin user updated:', adminId);
      return { success: true, admin: result.rows[0] };
    } catch (err) {
      console.error('Update admin user error:', err);
      return { success: false, error: err.message };
    }
  }

  // ── DELETE ADMIN USER ────────────────────────────────────────────────

  static async deleteAdminUser(adminId) {
    try {
      const result = await pool.query(
        'DELETE FROM admin_users WHERE id = $1 RETURNING id, email',
        [adminId]
      );

      if (result.rows.length === 0) {
        return { success: false, error: 'Admin user not found' };
      }

      console.log('✓ Admin user deleted:', result.rows[0].email);
      return { success: true, deleted: result.rows[0] };
    } catch (err) {
      console.error('Delete admin user error:', err);
      return { success: false, error: err.message };
    }
  }

  // ── GET ACTIVITY LOG ─────────────────────────────────────────────────

  static async getActivityLog(limit = 100, offset = 0, filters = {}) {
    try {
      let query = `SELECT aal.*, au.email as admin_email
                   FROM admin_activity_log aal
                   LEFT JOIN admin_users au ON aal.admin_user_id = au.id
                   WHERE 1=1`;
      const params = [];
      let paramCount = 1;

      if (filters.admin_id) {
        query += ` AND aal.admin_user_id = $${paramCount}`;
        params.push(filters.admin_id);
        paramCount++;
      }

      if (filters.action) {
        query += ` AND aal.action = $${paramCount}`;
        params.push(filters.action);
        paramCount++;
      }

      if (filters.resource_type) {
        query += ` AND aal.resource_type = $${paramCount}`;
        params.push(filters.resource_type);
        paramCount++;
      }

      query += ` ORDER BY aal.created_at DESC LIMIT $${paramCount} OFFSET $${paramCount + 1}`;
      params.push(limit, offset);

      const result = await pool.query(query, params);

      const totalResult = await pool.query('SELECT COUNT(*) as count FROM admin_activity_log');
      
      return {
        success: true,
        logs: result.rows,
        total: parseInt(totalResult.rows[0].count)
      };
    } catch (err) {
      console.error('Get activity log error:', err);
      return { success: false, error: err.message };
    }
  }

  // ── UPDATE ADMIN LAST LOGIN ──────────────────────────────────────────

  static async updateLastLogin(adminId) {
    try {
      await pool.query(
        'UPDATE admin_users SET last_login = NOW() WHERE id = $1',
        [adminId]
      );
    } catch (err) {
      console.error('Update last login error:', err);
    }
  }
}

module.exports = UserService;

File: fastapi_backend/services/user_service.py
import json
import bcrypt
from db import get_pool


async def get_all_companies(limit: int = 50, offset: int = 0) -> dict:
    pool = await get_pool()
    async with pool.acquire() as conn:
        rows = await conn.fetch(
            """SELECT c.id, c.name, c.status, c.active_modules, c.region, c.entity_name, c.created_at,
                      COUNT(DISTINCT u.id) as team_size,
                      MAX(u.email) FILTER (WHERE u.permission_level = 'owner') as owner_email
               FROM companies c
               LEFT JOIN users u ON u.company_id = c.id
               GROUP BY c.id
               ORDER BY c.created_at DESC LIMIT $1 OFFSET $2""",
            limit, offset,
        )
        total = await conn.fetchval("SELECT COUNT(*) FROM companies")
        return {"success": True, "companies": [dict(r) for r in rows], "total": total}


async def get_company(company_id: int) -> dict:
    pool = await get_pool()
    async with pool.acquire() as conn:
        row = await conn.fetchrow(
            """SELECT c.id, c.name, c.status, c.active_modules, c.region, c.entity_name, c.created_at,
                      COUNT(DISTINCT u.id) as team_size,
                      MAX(u.email) FILTER (WHERE u.permission_level = 'owner') as owner_email
               FROM companies c
               LEFT JOIN users u ON u.company_id = c.id
               WHERE c.id = $1 GROUP BY c.id""",
            company_id,
        )
        if not row:
            return {"success": False, "error": "Company not found"}
        return {"success": True, "company": dict(row)}


async def get_all_admin_users(limit: int = 50, offset: int = 0) -> dict:
    pool = await get_pool()
    async with pool.acquire() as conn:
        rows = await conn.fetch(
            """SELECT a.id, a.email, a.admin_level, a.company_id, a.permissions,
                      a.is_active, a.created_at, a.last_login,
                      c.name as client_name
               FROM admin_users a LEFT JOIN companies c ON a.company_id = c.id
               ORDER BY a.created_at DESC LIMIT $1 OFFSET $2""",
            limit, offset,
        )
        total = await conn.fetchval("SELECT COUNT(*) FROM admin_users")
        return {"success": True, "admins": [dict(r) for r in rows], "total": total}


async def create_admin_user(data: dict) -> dict:
    pool = await get_pool()
    email = data.get("email")
    password = data.get("password")
    admin_level = data.get("admin_level")
    company_id = data.get("company_id")
    permissions = data.get("permissions", {"pricing": True, "features": True, "users": True})

    if not email or not password or not admin_level:
        return {"success": False, "error": "Email, password, and admin_level required"}
    if admin_level not in ("super_admin", "admin"):
        return {"success": False, "error": "Invalid admin_level"}

    password_hash = bcrypt.hashpw(password.encode(), bcrypt.gensalt(12)).decode()

    async with pool.acquire() as conn:
        try:
            row = await conn.fetchrow(
                """INSERT INTO admin_users (email, password_hash, admin_level, company_id, permissions, is_active)
                   VALUES ($1,$2,$3,$4,$5,true)
                   RETURNING id, email, admin_level, company_id, permissions, is_active, created_at""",
                email, password_hash, admin_level, company_id, json.dumps(permissions),
            )
            return {"success": True, "admin": dict(row)}
        except Exception as e:
            if "23505" in str(e):
                return {"success": False, "error": "Email already exists"}
            return {"success": False, "error": str(e)}


async def update_admin_user(admin_id: int, updates: dict) -> dict:
    pool = await get_pool()
    allowed = ["is_active", "permissions"]
    fields, values = [], []
    p = 1
    for key, val in updates.items():
        if key in allowed:
            fields.append(f"{key} = ${p}")
            values.append(json.dumps(val) if key == "permissions" else val)
            p += 1
    if not fields:
        return {"success": False, "error": "No valid fields to update"}
    values.append(admin_id)
    async with pool.acquire() as conn:
        row = await conn.fetchrow(
            f"UPDATE admin_users SET {', '.join(fields)}, updated_at = NOW() WHERE id = ${p} RETURNING *", *values
        )
        if not row:
            return {"success": False, "error": "Admin user not found"}
        return {"success": True, "admin": dict(row)}


async def delete_admin_user(admin_id: int) -> dict:
    pool = await get_pool()
    async with pool.acquire() as conn:
        row = await conn.fetchrow("DELETE FROM admin_users WHERE id = $1 RETURNING id, email", admin_id)
        if not row:
            return {"success": False, "error": "Admin user not found"}
        return {"success": True, "deleted": dict(row)}


async def get_activity_log(limit: int = 100, offset: int = 0, filters: dict = {}) -> dict:
    pool = await get_pool()
    query = """SELECT aal.*, au.email as admin_email, u.email as user_email
               FROM admin_activity_log aal
               LEFT JOIN admin_users au ON aal.admin_user_id = au.id
               LEFT JOIN users u ON aal.user_id = u.id
               WHERE 1=1"""
    params = []
    p = 1
    if filters.get("admin_id"):
        query += f" AND aal.admin_user_id = ${p}"
        params.append(filters["admin_id"])
        p += 1
    if filters.get("action"):
        query += f" AND aal.action = ${p}"
        params.append(filters["action"])
        p += 1
    if filters.get("resource_type"):
        query += f" AND aal.resource_type = ${p}"
        params.append(filters["resource_type"])
        p += 1
    if filters.get("module_key"):
        query += f" AND aal.module_key = ${p}"
        params.append(filters["module_key"])
        p += 1
    if filters.get("company_id"):
        query += f" AND aal.company_id = ${p}"
        params.append(filters["company_id"])
        p += 1
    query += f" ORDER BY aal.created_at DESC LIMIT ${p} OFFSET ${p+1}"
    params.extend([limit, offset])
    async with pool.acquire() as conn:
        rows = await conn.fetch(query, *params)
        total = await conn.fetchval("SELECT COUNT(*) FROM admin_activity_log")
        return {"success": True, "logs": [dict(r) for r in rows], "total": total}


async def log_admin_activity(admin_user_id: int, action: str, resource_type: str, resource_id, changes=None):
    pool = await get_pool()
    async with pool.acquire() as conn:
        try:
            await conn.execute(
                """INSERT INTO admin_activity_log (admin_user_id, action, resource_type, resource_id, changes)
                   VALUES ($1,$2,$3,$4,$5)""",
                admin_user_id, action, resource_type, resource_id,
                json.dumps(changes) if changes else None,
            )
        except Exception as e:
            print("Log activity error:", e)


async def log_trust_activity(user_id: int, company_id: int, action: str, resource_type: str, resource_id, changes=None):
    """POMAR Trust's equivalent of log_admin_activity, for regular-user actors
    (site engineers, GC owners) rather than admin_users. Writes into the same
    admin_activity_log table/Admin Portal viewer via the widened actor columns
    (admin_activity_log.user_id/company_id/module_key) instead of a separate log."""
    pool = await get_pool()
    async with pool.acquire() as conn:
        try:
            await conn.execute(
                """INSERT INTO admin_activity_log (user_id, company_id, module_key, action, resource_type, resource_id, changes)
                   VALUES ($1,$2,'trust',$3,$4,$5,$6)""",
                user_id, company_id, action, resource_type, resource_id,
                json.dumps(changes) if changes else None,
            )
        except Exception as e:
            print("Log trust activity error:", e)
